Userland/Libraries/LibDisassembly/riscv64/Encoding.h

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibDisassembly/riscv64/Registers.h>
+
+namespace Disassembly::RISCV64 {
+
+struct RawRType {
+    u8 funct7;
+    Register rs2;
+    Register rs1;
+    u8 funct3;
+    Register rd;
+    u8 opcode;
+
+    static RawRType parse(u32 instruction);
+};
+struct RawIType {
+    i32 imm;
+    Register rs1;
+    u8 funct3;
+    Register rd;
+    u8 opcode;
+
+    static RawIType parse(u32 instruction);
+};
+struct RawSType {
+    i32 imm;
+    Register rs2;
+    Register rs1;
+    u8 funct3;
+    u8 opcode;
+
+    static RawSType parse(u32 instruction);
+};
+struct RawBType {
+    i32 imm;
+    Register rs2;
+    Register rs1;
+    u8 funct3;
+    u8 opcode;
+
+    static RawBType parse(u32 instruction);
+};
+struct RawUType {
+    i32 imm;
+    Register rd;
+    u8 opcode;
+
+    static RawUType parse(u32 instruction);
+};
+struct RawJType {
+    i32 imm;
+    Register rd;
+    u8 opcode;
+
+    static RawJType parse(u32 instruction);
+};
+struct RawR4Type {
+    Register rs3;
+    u8 fmt;
+    Register rs2;
+    Register rs1;
+    u8 funct3;
+    Register rd;
+    u8 opcode;
+
+    static RawR4Type parse(u32 instruction);
+};
+struct RawCRType {
+    u8 funct4;
+    Register rd_or_rs1;
+    Register rs2;
+    u8 opcode;
+
+    static RawCRType parse(u16 instruction);
+};
+struct RawCIType {
+    u8 funct3;
+    i32 imm;
+    Register rd_or_rs1;
+    u8 opcode;
+
+    static RawCIType parse(u16 instruction);
+};
+struct RawCSSType {
+    u8 funct3;
+    i32 imm;
+    Register rs2;
+    u8 opcode;
+
+    static RawCSSType parse(u16 instruction);
+};
+struct RawCIWType {
+    u8 funct3;
+    i32 imm;
+    Register rd;
+    u8 opcode;
+
+    static RawCIWType parse(u16 instruction);
+};
+struct RawCLType {
+    u8 funct3;
+    i32 imm;
+    Register rs1;
+    Register rd;
+    u8 opcode;
+
+    static RawCLType parse(u16 instruction);
+};
+struct RawCSType {
+    u8 funct3;
+    i32 imm;
+    Register rs1;
+    Register rs2;
+    u8 opcode;
+
+    static RawCSType parse(u16 instruction);
+};
+struct RawCAType {
+    u8 funct6;
+    Register rs1_or_rd;
+    u8 funct2;
+    Register rs2;
+    u8 opcode;
+
+    static RawCAType parse(u16 instruction);
+};
+struct RawCBType {
+    u8 funct3;
+    i32 offset;
+    u8 funct2;
+    Register rs1;
+    u8 opcode;
+
+    static RawCBType parse(u16 instruction);
+};
+struct RawCJType {
+    u8 funct3;
+    i32 jump_target;
+    u8 opcode;
+
+    static RawCJType parse(u16 instruction);
+};
+
+// Table 24.1
+enum class MajorOpcode : u8 {
+    LOAD = 0b00000'11,
```

Comment:
Odd `'` placement, but theres likely a reason?

Reply:
Yup: the lowest 2 bits are always 1 and specify the instruction size. This is similar to how they represent it in the spec and the table named here.